src/Sarif.Viewer.VisualStudio.UnitTests/Sarif/StackFrameExtensionsTests.cs

```diff
@@ -0,0 +1,99 @@
+            int runIndex = 0;
+            string moduleName = "Test.Module";
+            var stackFrame = new StackFrame
+            {
+                Module = moduleName,
+            };
+
+            var stackFrameModel = stackFrame.ToStackFrameModel(resultId, runIndex);
+            stackFrameModel.Should().NotBeNull();
+            stackFrameModel.ResultId.Should().Be(resultId);
+            stackFrameModel.RunIndex.Should().Be(runIndex);
+            stackFrameModel.Module.Should().BeEquivalentTo(moduleName);
+            stackFrameModel.Message.Should().BeNull();
+            stackFrameModel.FullyQualifiedLogicalName.Should().BeNull();
+            stackFrameModel.FilePath.Should().BeNull();
+            stackFrameModel.Region.Should().BeNull();
+            stackFrameModel.Address.Should().Be(0);
+            stackFrameModel.Line.Should().Be(0);
+            stackFrameModel.Column.Should().Be(0);
+            stackFrameModel.Offset.Should().Be(0);
+        }
+
+        [Fact]
+        public void StackFrameExtensions_HasLocation()
+        {
+            int resultId = 1;
+            int runIndex = 0;
+            var stackFrame = new StackFrame
+            {
+                Location = new Location
+                {
+                    Message = new Message { Text = "stack frame location 1" },
+                    LogicalLocation = new LogicalLocation
+                    {
+                        FullyQualifiedName = @"\root\FQDN\path-to\file.ext",
+                    },
+                    PhysicalLocation = new PhysicalLocation
+                    {
+                        Address = new Address
+                        {
+                            AbsoluteAddress = 0xFF,
+                            OffsetFromParent = 128,
+                        },
+                        ArtifactLocation = new ArtifactLocation
+                        {
+                            Uri = new Uri("path/to/file.cpp", UriKind.Relative),
+                        },
+                        Region = new Region
+                        {
+                            StartLine = 10,
+                            StartColumn = 5,
+                        },
+                    }
+                }
+            };
+
+            var stackFrameModel = stackFrame.ToStackFrameModel(resultId, runIndex);
+            stackFrameModel.Should().NotBeNull();
+            stackFrameModel.ResultId.Should().Be(resultId);
+            stackFrameModel.RunIndex.Should().Be(runIndex);
+            stackFrameModel.Module.Should().BeNull();
+            stackFrameModel.Message.Should().BeEquivalentTo("stack frame location 1");
+            stackFrameModel.FullyQualifiedLogicalName.Should().BeEquivalentTo(@"\root\FQDN\path-to\file.ext");
+            stackFrameModel.FilePath.Should().BeEquivalentTo("path/to/file.cpp");
+            stackFrameModel.Region.Should().NotBeNull();
+            stackFrameModel.Region.StartLine.Should().Be(10);
```

Comment:
>StartLine [](http://example.com/codeflow?start=35&length=9)

This test code is not very maintainable, as it creates multiple copies of literals. You could have cleaned this up to some degree by consulting values from the stackFrame object itself, which you retained. 

stackFrameModel.Region.Startlein.Should().Be(stackFrame.Location.PhysicalLocation.Region.StartLine).

In general, when you find yourself repeating hard-coded strings and values, like your relative path @"\root\", etc., that's a clue to try to reduce the code to a single copy, i.e., authoring the code where the literal value only appears once.

This is a basic quality standard that we enforce through-out our codebase (and which is very common in many projects).

Reply:
thanks for the feedback, refactor the test case and move the verification into a common function which does not depends on hardcoded values. please review pr: [https://github.com/microsoft/sarif-visualstudio-extension/pull/448](https://github.com/microsoft/sarif-visualstudio-extension/pull/448)